### PR TITLE
Elasticsearch: Enable logs samples for metric queries

### LINKED ISSUE #62212: Elasticsearch: add support to logs sample
the logs-sample functionality was added in https://github.com/grafana/grafana/pull/61864 for the `loki` datasource.
we should add support to the `elasticsearch` datasource too.

currently, when you visit in explore a newly created elasticseach datasource, you automatically see some data: a graph.
with logs-samples you would also see some log-lines there, making the page a lot more useful for getting an initial understanding of what can be on the page.

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**

this enables log samples for elasticsearch metric queries

**Which issue(s) does this PR fix?**:

Fixes #62212

**Special notes for your reviewer:**

https://github.com/grafana/grafana/assets/34524710/27f15647-b2b5-42e8-8009-0d1ceb525b04

### CODE REVIEW

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 1]
```diff
@@ -534,13 +536,15 @@ export class ElasticDatasource
...
         return undefined;
     }
   }
 
   getSupportedSupplementaryQueryTypes(): SupplementaryQueryType[] {
-    return [SupplementaryQueryType.LogsVolume];
+    return [SupplementaryQueryType.LogsVolume, SupplementaryQueryType.LogsSample];
   }
 
   getSupplementaryQuery(options: SupplementaryQueryOptions, query: ElasticsearchQuery): ElasticsearchQuery | undefined {
```
Comment (matyax) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:550]: Could you please update `elasticsearch/datasource.test.ts` covering the code added in this PR? Thanks! 

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/elasticsearch/datasource.test.ts#L926

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 2]
```diff
@@ -593,6 +597,24 @@ export class ElasticDatasource
           bucketAggs,
         };
 
+      case SupplementaryQueryType.LogsSample:
+        isQuerySuitable = !!(
+          query.metrics?.length === 1 &&
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:602]: We should show logs sample only for `date_histogram` queries, which is a type of metric queries.

**Review by ivanahuckova — COMMENTED** [review]
Left a comment bellow. Moreover, for `date_histogram` queries we should add prefered viz type to `graph` (both in frontend and backend). Doesn't have to be part of this PR, but should be a next step. 🙂

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 3]
```diff
@@ -593,6 +597,21 @@ export class ElasticDatasource
...
+        );
+
+        if (!isQuerySuitable) {
+          return undefined;
+        }
+
+        return {
+          refId: `${REF_ID_STARTER_LOG_SAMPLE}${query.refId}`,
+          query: query.query,
+          metrics: [{ type: 'logs', id: '1' }],
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:612]: Here you are not using the `limit` that you are setting to `100`.

Comment (ivanahuckova) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:612]: Could you also add test here when you fix it, that the limit is applied

**Review by ivanahuckova — COMMENTED** [review]
Left couple more comments. 🙂 I am so happy to see logs sample in Elasticsearch 🚀 .

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/index.tsx [hunk 4]
```diff
@@ -93,7 +93,7 @@ const QueryEditorForm = ({ value }: Props) => {
   const styles = useStyles2(getStyles);
 
   // To be considered a time series query, the last bucked aggregation must be a Date Histogram
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/index.tsx:95]: I think that this comment should be moved to `isTimeSeriesQuery`

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 5]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
         timeField: '',
       });
     });
+
+    it('does not return logs samples for non date_histogram queries', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:982]:
```suggestion
    it('does not return logs samples for non time series queries', () => {
```

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 6]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
...
+          {
+            refId: 'A',
+            bucketAggs: [{ type: 'filters', id: '1' }],
+            query: '',
+          }
+        )
+      ).toEqual(undefined);
+    });
+
+    it('returns logs samples for data_histogram queries', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:995]:
```suggestion
    it('returns logs samples for time series queries', () => {
```

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 7]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
...
+    });
+  });
+
+  describe('getDataProvider', () => {
+    let ds: ElasticDatasource;
+    beforeEach(() => {
+      ds = getTestContext().ds;
+    });
+
+    it('does not create a logs sample provider for logs query', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:1019]:
```suggestion
    it('does not create a logs sample provider for non time series query', () => {
```

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 8]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
...
+            refId: 'A',
+            metrics: [{ type: 'logs', id: '1', settings: { limit: '100' } }],
+          },
+        ],
+      });
+
+      expect(ds.getDataProvider(SupplementaryQueryType.LogsSample, options)).not.toBeDefined();
+    });
+
+    it('does create a logs sample provider for date_histogram query', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:1032]:
```suggestion
    it('does create a logs sample provider for time series query', () => {
```

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 9]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
...
+    });
+  });
+
+  describe('getLogsSampleDataProvider', () => {
+    let ds: ElasticDatasource;
+    beforeEach(() => {
+      ds = getTestContext().ds;
+    });
+
+    it("doesn't return a logs sample provider given a log query", () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:1052]:
```suggestion
    it("doesn't return a logs sample provider given anon time series query", () => {
```

File: public/app/plugins/datasource/elasticsearch/datasource.test.ts [hunk 10]
```diff
@@ -977,6 +978,102 @@ describe('ElasticDatasource', () => {
...
+            refId: 'A',
+            metrics: [{ type: 'logs', id: '1', settings: { limit: '100' } }],
+          },
+        ],
+      });
+
+      expect(ds.getLogsSampleDataProvider(request)).not.toBeDefined();
+    });
+
+    it('returns a logs sample provider given a date_histogram query', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.test.ts:1065]:
```suggestion
    it('returns a logs sample provider given a time series query', () => {
```

File: public/app/plugins/datasource/elasticsearch/utils.ts [hunk 11]
```diff
@@ -101,3 +101,7 @@ export const isSupportedVersion = (version: SemVer): boolean => {
 
 export const unsupportedVersionMessage =
   'Support for Elasticsearch versions after their end-of-life (currently versions < 7.16) was removed. Using unsupported version of Elasticsearch may lead to unexpected and incorrect results.';
+
+export const isTimeSeriesQuery = (query: ElasticsearchQuery): boolean => {
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/plugins/datasource/elasticsearch/utils.ts:105]: Last thing - could you please add some tests for this - such as:
- returns false when log query
- returns false when empty bucketAggs
- returns false when empty date_histogram is not last
- returns truw when empty date_histogram is last

and any other test you can come up with 🙂

**Review by ivanahuckova — COMMENTED** [review]
Left some comments bellow, but looks good!

**Review by ivanahuckova — APPROVED** [review]
Great job and :bufo-yay: for having logs sample for Es.

**Pull request merged**